File: src/main/java/com/everyonewaiter/global/logging/DiscordLogbackAppender.java
package com.everyonewaiter.global.logging;

import static com.everyonewaiter.domain.support.DateFormatter.SERIALIZE;

import ch.qos.logback.classic.spi.ILoggingEvent;
import ch.qos.logback.classic.spi.ThrowableProxyUtil;
import ch.qos.logback.core.UnsynchronizedAppenderBase;
import com.everyonewaiter.domain.notification.DiscordColor;
import com.everyonewaiter.domain.notification.DiscordEmbed;
import com.everyonewaiter.domain.notification.DiscordEmbeds;
import com.everyonewaiter.domain.notification.DiscordField;
import java.io.IOException;
import java.time.LocalDateTime;
import java.util.Map;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpRequest;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.MediaType;
import org.springframework.http.client.ClientHttpResponse;
import org.springframework.web.client.RestClient;

@Setter
@RequiredArgsConstructor
public class DiscordLogbackAppender extends UnsynchronizedAppenderBase<ILoggingEvent> {

  private static final Logger LOGGER = LoggerFactory.getLogger(DiscordLogbackAppender.class);
  private static final RestClient WEB_HOOK_CLIENT =
      RestClient.create("https://discord.com/api/webhooks");

  private String discordWebhookUri;

  @Override
  protected void append(ILoggingEvent iLoggingEvent) {
    DiscordEmbeds embeds = createEmbeds(iLoggingEvent);

    WEB_HOOK_CLIENT.post()
        .uri("/" + discordWebhookUri)
        .contentType(MediaType.APPLICATION_JSON)
        .body(embeds)
        .retrieve()
        .onStatus(HttpStatusCode::is4xxClientError, this::handleError)
        .onStatus(HttpStatusCode::is5xxServerError, this::handleError)
        .toBodilessEntity();
  }

  private DiscordEmbeds createEmbeds(ILoggingEvent iLoggingEvent) {
    Map<String, String> mdc = iLoggingEvent.getMDCPropertyMap();
    String throwable = ThrowableProxyUtil.asString(iLoggingEvent.getThrowableProxy());
    String stackTrace = throwable.length() > 600 ? throwable.substring(0, 600) : throwable;

    DiscordEmbed errorInfoEmbed = new DiscordEmbed(
        "Error Information",
        iLoggingEvent.getFormattedMessage(),
        DiscordColor.RED
    );
    errorInfoEmbed.addField(new DiscordField("Timestamp", SERIALIZE.format(LocalDateTime.now())));
    errorInfoEmbed.addField(new DiscordField("Request URI", mdc.get("requestURI")));
    errorInfoEmbed.addField(new DiscordField("Request Parameters", mdc.get("requestParameters")));
    errorInfoEmbed.addField(new DiscordField("Request Headers", mdc.get("requestHeaders")));
    errorInfoEmbed.addField(new DiscordField("Request Cookies", mdc.get("requestCookies")));

    DiscordEmbed stackTraceEmbed = new DiscordEmbed(
        "Stack Trace",
        "```java\n" + stackTrace + "\n```",
        DiscordColor.RED
    );

    return new DiscordEmbeds(errorInfoEmbed, stackTraceEmbed);
  }

  private void handleError(
      HttpRequest request,
      ClientHttpResponse response
  ) throws IOException {
    String responseBody = new String(response.getBody().readAllBytes());
    LOGGER.warn("[{} 디스코드 메시지 전송 실패] {}", response.getStatusText(), responseBody);
  }

}

File: src/main/kotlin/com/everyonewaiter/global/logging/DiscordLogbackAppender.kt
package com.everyonewaiter.global.logging

import ch.qos.logback.classic.spi.ILoggingEvent
import ch.qos.logback.classic.spi.ThrowableProxyUtil
import ch.qos.logback.core.UnsynchronizedAppenderBase
import com.everyonewaiter.common.notification.discord.DiscordEmbed
import com.everyonewaiter.common.notification.discord.DiscordField
import com.everyonewaiter.common.notification.discord.DiscordWebhookPayload
import io.github.oshai.kotlinlogging.KotlinLogging
import org.springframework.http.HttpRequest
import org.springframework.http.HttpStatusCode
import org.springframework.http.MediaType
import org.springframework.http.client.ClientHttpResponse
import org.springframework.web.client.RestClient
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

private const val COLOR_RED = 16711680

private val logger = KotlinLogging.logger {}

class DiscordLogbackAppender : UnsynchronizedAppenderBase<ILoggingEvent>() {
    private val discordWebhookClient = RestClient.create("https://discord.com/api/webhooks")

    @Suppress("MemberVisibilityCanBePrivate")
    lateinit var discordWebhookUri: String

    override fun append(loggingEvent: ILoggingEvent) {
        val payload = createPayload(loggingEvent)
        discordWebhookClient
            .post()
            .uri("/$discordWebhookUri")
            .contentType(MediaType.APPLICATION_JSON)
            .body(payload)
            .retrieve()
            .onStatus(HttpStatusCode::is4xxClientError, this::handleError)
            .onStatus(HttpStatusCode::is5xxServerError, this::handleError)
            .toBodilessEntity()
    }

    private fun createPayload(loggingEvent: ILoggingEvent): DiscordWebhookPayload {
        val embeds = mutableListOf<DiscordEmbed>()
        val mdc: Map<String, String> = loggingEvent.mdcPropertyMap
        val throwable = ThrowableProxyUtil.asString(loggingEvent.throwableProxy)
        val stackTrace = if (throwable.length > 700) throwable.substring(0, 700) else throwable

        embeds.add(
            DiscordEmbed(
                title = "Error Information",
                description = loggingEvent.formattedMessage,
                color = COLOR_RED,
                fields = listOf(
                    DiscordField(
                        name = "Timestamp",
                        value = LocalDateTime.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")),
                    ),
                    DiscordField(
                        name = "Request ID",
                        value = mdc["requestId"] ?: "",
                    ),
                    DiscordField(
                        name = "Request URI",
                        value = mdc["requestUri"] ?: "",
                    ),
                    DiscordField(
                        name = "Request Parameters",
                        value = mdc["requestParameters"] ?: "",
                    ),
                    DiscordField(
                        name = "Request Headers",
                        value = mdc["requestHeaders"] ?: "",
                    ),
                    DiscordField(
                        name = "Request Cookies",
                        value = mdc["requestCookies"] ?: "",
                    ),
                ),
            ),
        )
        embeds.add(
            DiscordEmbed(
                title = "Stack Trace",
                description = "```kotlin\n$stackTrace\n```",
                color = COLOR_RED,
                fields = listOf(),
            ),
        )

        return DiscordWebhookPayload(embeds)
    }

    @Suppress("UNUSED_PARAMETER")
    private fun handleError(
        request: HttpRequest,
        response: ClientHttpResponse,
    ) {
        logger.warn { "[디스코드 메시지 전송 실패] ${response.statusText} : ${String(response.body.readAllBytes())}" }
    }
}
